### PR TITLE
add hiprtc APIs

### PR DESCRIPTION
**Pull request by weihanmines**
Thanks for your contribution! Unfortunately, tensorflow/runtime is currently not
accepting contributions. Please see the
[Contribution Guidelines](../blob/master/README.md#contribution-guidelines) for
more information.

This PR only contains the changes needed for the hip just-in-time compilation (hiprtc). The hiprtc headers and implementation are put at the end of hip APIs (see generate.sh) because hiprtc implementation is rolled up in libamdhip64.so which is the same as hip APIs. New PR will be created once hiprtc is separated from hip.  

A command line option is added to generate.sh. Bazel requires the option.

@chsigg @hanbinyoon

### CODE REVIEW

File: backends/gpu/tools/stub_codegen/generate.sh [hunk 1]
```diff
@@ -34,3 +39,10 @@ for API in "hip" "rocblas" "rocsolver" "miopen" "hipfft"; do
    ./bazel-bin/backends/gpu/tools/stub_codegen/impl_codegen \
        $(dirname $0)/$API.json | clang-format > $(printf $SRC_PATH $API)
 done
+
+# Hiprtc is currently rolled up in hip shared library.
```
Comment (chsigg) [inline backends/gpu/tools/stub_codegen/generate.sh:43]: If it's going to be split up, would it make sense to already have a separate hiprtc_stub? It should be OK for two `LoadSymbol` functions to load the same shared object file. `llvm::sys::DynamicLibrary::getPermanentLibrary()` says "It is safe to call this function multiple times for the same library."

**Review by chsigg — COMMENTED** [review]
It seems OK to add this, but I think we came to the conclusion that we don't really need this and would instead just use HSACO in the tests? Is this a problem because HSACO is architecture specific, and you want to be able to run the same test across multiple architectures?

File: backends/gpu/tools/stub_codegen/generate.sh (hunk 1 above)
Comment (weihanmines) [reply backends/gpu/tools/stub_codegen/generate.sh:43]: > If it's going to be split up, would it make sense to already have a separate hiprtc_stub? It should be OK for two `LoadSymbol` functions to load the same shared object file. `llvm::sys::DynamicLibrary::getPermanentLibrary()` says "It is safe to call this function multiple times for the same library."

Thanks. I will split it up.

**Comment by weihanmines** [comment]
> It seems OK to add this, but I think we came to the conclusion that we don't really need this and would instead just use HSACO in the tests? Is this a problem because HSACO is architecture specific, and you want to be able to run the same test across multiple architectures?

I haven't do experiments using HSACO. Let me do some experiments. I got hiprtc working so far. It is the reason why I did not change the ModuleLoadData expression in driver_wrapper.cc and push out the module and launch mlir files.

File: backends/gpu/include/tfrt/gpu/wrapper/hiprtc_wrapper.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                                const char* name,
+                                int numberHeaders,
+                                const char** headers,
+                                const char** includeNames);
+llvm::Error HiprtcDestroyProgram(hiprtcProgram* prog);
+llvm::Error HiprtcGetLoweredName(
+                                 hiprtcProgram prog,
+                                 const char* name_expression,
+                                 const char** lowered_name);
+llvm::Error HiprtcGetProgramLog(hiprtcProgram prog, char* log);
```
[the remark this answers is not in the document]
Comment (weihanmines) [reply backends/gpu/include/tfrt/gpu/wrapper/hiprtc_wrapper.h:49]: I could not find any official documentation on hiprtc. I use the header file for reference. I will let you know once I find the official document.

[the remark this answers is not in the document]
Comment (chsigg) [reply backends/gpu/include/tfrt/gpu/wrapper/hiprtc_wrapper.h:49]: That's fine. The API is mostly self-explanatory.

**Comment by weihanmines** [comment]
Hi Christian @chsigg, I have removed the duplication from the headers. Let me know if there is anything else I could do to help the merge.

**Pull request merged**